### DISCUSSION TITLE
iOS Mobile

### DISCUSSION THREAD

**Discussion by bharatmalhotra84** [vladmandic/human#265] [Q&A] [answered]
Hi,

I'm using the libary and can see it has troubles showing the face dectection frame on iOS (mobile). After allowing the camera, it does not load the frame. Is there anything I should be looking at specifically to make it work?

Your input is much appreciated.

Regards,
Bharat

**Comment by vladmandic** [comment]
without providing more information, i cannot really help.  

you haven't even stated are you using safari or chrome?
i'd suggest to try with chrome and if it doesn't work, attach chrome inspector to instance running on ios so you can monitor browser console output.

note that ios is supported as best-effort as i do not have ios device available for testing.

**Reply by bharatmalhotra84** [reply]
Hi Vlad,

Sorry I did not mention it. I'm checking it on safari (iphone 12). I can see console errors using web inspector. see the attached.

![MicrosoftTeams-image](https://user-images.githubusercontent.com/31948676/162979662-762bacce-1b8c-4982-9348-9e9fa683563e.png)

**Reply by bharatmalhotra84** [reply]
I tried cleaning up cache on my safari but no luck.

**Comment by vladmandic** [comment, marked as answer]
first exception is in your code, not inside `human` library - **jQuery null is not an object**.  
there is exception at the end as well, `NotAllowedErrror`, but cant see exactly where. check between browser and web server if settings are correct to use things like camera access.

**Reply by bharatmalhotra84** [reply]
Hi Vlad, sorry for the delay reaching on this. I have resolved all other errors except this one. I was trying to reach out to other platforms to separate the issue from the library and iphone (safari).
![image](https://user-images.githubusercontent.com/31948676/164471072-0853812d-f1e4-4bcc-9d24-871a22f5be9e.png)

**Reply by bharatmalhotra84** [reply]
I checked safari camera settings and it's allowed

**Comment by vladmandic** [comment]
it would help if you could copy&paste errors instead of hard to read screenshots.
but important part - expand that exception so it shows more details - i don't know what is in you `main.js` line 95

in general, this is most commonly related to browser enforcing different things when working with video streams such as:
- when requesting webcam video stream, make sure you set `audio: none` or browser may block access  
- try calling `video.play()` in some `onclick` event instead?  
  a lot of browsers don't allow autoplay of videos unless there is some user interaction

**Reply by bharatmalhotra84** [reply]
ok, I will try to pull some internal information here. I see audio is set to false. do you suggest  changes in below code so video starts on a some kind of user event instead of autoplay?
async function webCam() {
    try {

        status("starting webcam...");
        const options = {
            audio: false,
            video: {facingMode: "user", resizeMode: "none", width: {ideal: document.body.clientWidth}}
        };
        const stream = await navigator.mediaDevices.getUserMedia(options);
        const ready = new Promise((resolve) => {
            dom.video.onloadeddata = () => resolve(true);
        });
        dom.video.srcObject = stream;
        dom.video.play();
        await ready;
        dom.canvas.width = dom.video.videoWidth;
        dom.canvas.height = dom.video.videoHeight;
        const track = stream.getVideoTracks()[0];
        const capabilities = track.getCapabilities ? track.getCapabilities() : "";
        const settings = track.getSettings ? track.getSettings() : "";
        const constraints = track.getConstraints ? track.getConstraints() : "";
    }
    catch (e) {
        status("Problem starting webcam, refresh page and make sure you allow your browser access.");
    }
}

**Reply by bharatmalhotra84** [reply]
I can see also in the html, I added <video id="video" class="flipped" autoplay muted></video> so i will try removing autoplay

**Comment by vladmandic** [comment]
thats already in demo already, just first remove `dom.video.play()` so it doesnt try to autoplay from webcam init method
and then use something like this:
<https://github.com/vladmandic/human/blob/8aa4d3d6472011d7968bf9264fe0285ded18bd9e/demo/typescript/index.ts#L65-L68>

**Reply by bharatmalhotra84** [reply]
Thanks Vlad. I also sent you another code that is actually responsible for autoplay. If i remove it the video does not load at all
![image](https://user-images.githubusercontent.com/31948676/164711926-ad36c1ff-7691-4209-b280-22743f5d630c.png)

**Comment by vladmandic** [comment]
does it not load or does not start playing? if you're using canvas overlay over video remember that first frame does not get rendered until its processed, so it shows as blank even if video is loaded. i'm basically guessing here since i don't know your entire code.

**Reply by bharatmalhotra84** [reply]
I think you are right here. our requirement is video to be autoplay as soon as user is reached to this step. it works fine on all other browsers laptop or mobile except iphone safari v 15.4 / chrome / edge. if you have anything specific for this case, that would be great!

**Comment by vladmandic** [comment]
quick google shows some hints that video tag should have "autoplay muted playsinline" to work on newer iphones  
other than that, i don't have iphone at hand to try out anything

**Reply by bharatmalhotra84** [reply]
i have autoplay muted already and will try with playsinline and let you know.

**Reply by bharatmalhotra84** [reply]
adding playsinline did the trick! Thank you Vlad for your time and suggestions. have a nice weeked.
